cpu/nrf5x_common/periph/gpio.c

```diff
@@ -125,44 +145,77 @@ void gpio_write(gpio_t pin, int value)
 int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
                   gpio_cb_t cb, void *arg)
 {
-    /* disable external interrupt in case one is active */
-    NRF_GPIOTE->INTENSET &= ~(GPIOTE_INTENSET_IN0_Msk);
+    uint8_t _pin_index = 0xff;
+    /* Looking for already known pin in exti table */
+    for (unsigned int i = 0; i < _gpiote_next_index; i++) {
+        if (_exti_pins[i] == pin) {
+            _pin_index = i;
+            break;
+        }
+    }
+
+    /* New pin */
+    if (_pin_index == 0xff) {
+        assert(_gpiote_next_index < GPIOTE_CHAN_NUMOF);
+        _pin_index = _gpiote_next_index;
+        /* associate the current pin with channel index */
+        _exti_pins[_pin_index] = pin;
+        /* increase next index for next pin initialization */
+        _gpiote_next_index++;
+    }
+
     /* save callback */
-    exti_chan.cb = cb;
-    exti_chan.arg = arg;
+    exti_chan[_pin_index].cb = cb;
+    exti_chan[_pin_index].arg = arg;
     /* configure pin as input */
     gpio_init(pin, mode);
     /* set interrupt priority and enable global GPIOTE interrupt */
     NVIC_EnableIRQ(GPIOTE_IRQn);
     /* configure the GPIOTE channel: set even mode, pin and active flank */
-    NRF_GPIOTE->CONFIG[0] = (GPIOTE_CONFIG_MODE_Event |
+    NRF_GPIOTE->CONFIG[_pin_index] = (GPIOTE_CONFIG_MODE_Event |
                              (pin_num(pin) << GPIOTE_CONFIG_PSEL_Pos) |
 #ifdef CPU_MODEL_NRF52840XXAA
                              ((pin & PORT_BIT) << 8) |
 #endif
                              (flank << GPIOTE_CONFIG_POLARITY_Pos));
     /* enable external interrupt */
-    NRF_GPIOTE->INTENSET |= GPIOTE_INTENSET_IN0_Msk;
+    NRF_GPIOTE->INTENSET |= (GPIOTE_INTENSET_IN0_Msk << _pin_index);
+
     return 0;
 }
 
 void gpio_irq_enable(gpio_t pin)
 {
-    (void) pin;
-    NRF_GPIOTE->INTENSET |= GPIOTE_INTENSET_IN0_Msk;
+    for (unsigned int i = 0; i < _gpiote_next_index; i++) {
+        if (_exti_pins[i] == pin) {
+            NRF_GPIOTE->INTENSET |= (GPIOTE_INTENSET_IN0_Msk << i);
+            break;
+        }
+    }
 }
 
 void gpio_irq_disable(gpio_t pin)
 {
-    (void) pin;
-    NRF_GPIOTE->INTENCLR |= GPIOTE_INTENSET_IN0_Msk;
+    for (unsigned int i = 0; i < _gpiote_next_index; i++) {
+        if (_exti_pins[i] == pin) {
+            uint32_t inten = NRF_GPIOTE->INTENCLR;
+            NRF_GPIOTE->INTENCLR = (GPIOTE_INTENCLR_IN0_Msk << i);
+            inten &= ~(GPIOTE_INTENSET_IN0_Msk << i);
+            NRF_GPIOTE->INTENSET = inten;
```

Comment:
While testing I noticed that this is not needed. Just by doing ` NRF_GPIOTE->INTENCLR = (GPIOTE_INTENCLR_IN0_Msk << i);` the correspondent bit gets to 0.

Reply:
You are right, I re-tested without and it is still working. I tried many things yesterday before I could have a working version.

See also the last commit which contains another change: pressing a button with disabled interrupt was triggering the interrupt when re-enabling it. I fixed by clearing any pending interrupt when re-enabling. This is weird.

Reply:
> pressing a button with disabled interrupt was triggering the interrupt when re-enabling it. I fixed by clearing any pending interrupt when re-enabling. This is weird.

It seems that `INTENCLR` only disables interrupts from happening but not events from being registered. I changed that behavior by setting the `mode` bits of the `config` register to 'disabled', and setting them back to 'event' when re-enabling the irq.

I don't know if that is the correct way, in any case your solution only needs 1 line so I think that's fine.